File: src/main/java/me/miki/shindo/ui/comp/impl/CompToggleButton.java
package me.miki.shindo.ui.comp.impl;

import lombok.Getter;
import me.miki.shindo.management.color.AccentColor;
import me.miki.shindo.management.color.palette.ColorPalette;
import me.miki.shindo.management.color.palette.ColorType;
import me.miki.shindo.management.nanovg.NanoVGManager;
import me.miki.shindo.management.settings.impl.BooleanSetting;
import me.miki.shindo.ui.comp.Comp;
import me.miki.shindo.ui.framework.UIContext;
import me.miki.shindo.utils.ColorUtils;
import me.miki.shindo.utils.animation.ColorAnimation;
import me.miki.shindo.utils.animation.simple.SimpleAnimation;
import me.miki.shindo.utils.mouse.MouseUtils;

import java.awt.*;

public class CompToggleButton extends Comp {

    private final SimpleAnimation opacityAnimation = new SimpleAnimation();
    private final SimpleAnimation toggleAnimation = new SimpleAnimation();
    private final ColorAnimation circleAnimation = new ColorAnimation();

    @Getter
    private final BooleanSetting setting;

    @Getter
    private float scale;

    public CompToggleButton(float x, float y, float scale, BooleanSetting setting) {
        super(x, y);

        this.setting = setting;
        setScale(scale);
        toggleAnimation.setValue(setting.isToggled() ? 20.5F : 2.5F);
        circleAnimation.setColor(setting.isToggled() ? Color.WHITE : ctx().palette().getBackgroundColor(ColorType.DARK));
    }

    public CompToggleButton(BooleanSetting setting) {
        super(0, 0);

        this.setting = setting;
        setScale(1.0F);
        toggleAnimation.setValue(setting.isToggled() ? 20.5F : 2.5F);
        circleAnimation.setColor(setting.isToggled() ? Color.WHITE : ctx().palette().getBackgroundColor(ColorType.DARK));
    }

    public void setScale(float scale) {
        this.scale = scale;
        super.setWidth(34F * scale);
        super.setHeight(16F * scale);
    }

    @Override
    public void draw(int mouseX, int mouseY, float partialTicks) {

        UIContext ctx = ctx();
        NanoVGManager nvg = ctx.nvg();
        AccentColor accentColor = ctx.accent();
        ColorPalette palette = ctx.palette();

        float x = this.getX();
        float y = this.getY();
        float width = getWidth();
        float height = getHeight();
        float circle = 11 * scale;
        boolean toggled = setting.isToggled();

        opacityAnimation.setAnimation(toggled ? 1.0F : 0.0F, 14);
        toggleAnimation.setAnimation(toggled ? 20.5F : 2.5F, 14);

        nvg.drawRoundedRect(x, y, width, height, (7 * scale), palette.getBackgroundColor(ColorType.NORMAL));
        nvg.drawGradientRoundedRect(x, y, width, height, (7.4F * scale), ColorUtils.applyAlpha(accentColor.getColor1(), (int) (opacityAnimation.getValue() * 255)), ColorUtils.applyAlpha(accentColor.getColor2(), (int) (opacityAnimation.getValue() * 255)));
        nvg.drawRoundedRect(x + (toggleAnimation.getValue() * scale), y + (2.5F * scale), circle, circle, circle / 2, circleAnimation.getColor(toggled ? Color.WHITE : palette.getBackgroundColor(ColorType.DARK), 16));

        super.draw(mouseX, mouseY, partialTicks);
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {

        float x = this.getX();
        float y = this.getY();
        float width = getWidth();
        float height = getHeight();

        if (MouseUtils.isInside(mouseX, mouseY, x, y, width, height) && mouseButton == 0) {
            setting.setToggled(!setting.isToggled());
        }

        super.mouseClicked(mouseX, mouseY, mouseButton);
    }

}

File: src/main/kotlin/me/miki/shindo/ui/comp/impl/CompToggleButton.kt
package me.miki.shindo.ui.comp.impl

import me.miki.shindo.management.color.palette.ColorType
import me.miki.shindo.management.settings.impl.BooleanSetting
import me.miki.shindo.ui.comp.Comp
import me.miki.shindo.utils.ColorUtils
import me.miki.shindo.utils.animation.ColorAnimation
import me.miki.shindo.utils.animation.simple.SimpleAnimation
import me.miki.shindo.utils.mouse.MouseUtils
import java.awt.Color

class CompToggleButton : Comp {
    private val opacityAnimation = SimpleAnimation()
    private val toggleAnimation = SimpleAnimation()
    private val circleAnimation = ColorAnimation()

    private val setting: BooleanSetting
    private var scale: Float = 1.0f
    
    
    fun getSetting(): BooleanSetting = setting
    fun getScale(): Float = scale 

    constructor(x: Float, y: Float, scale: Float, setting: BooleanSetting) : super(x, y) {
        this.setting = setting
        
        setScale(scale)
        initState()
        
        
    }

    constructor(setting: BooleanSetting) : super(0f, 0f) {
        this.setting = setting
        setScale(1.0f)
        initState()
    }

    private fun initState() {
        toggleAnimation.value = if (setting.isToggled()) 20.5f else 2.5f
        circleAnimation.setColor(
            if (setting.isToggled()) Color.WHITE else palette.getBackgroundColor(ColorType.DARK)
        )
    }

    fun setScale(scale: Float) {
        this.scale = scale;
        super.setWidth(34F * scale);
        super.setHeight(16F * scale);
    }

    override fun draw(mouseX: Int, mouseY: Int, partialTicks: Float) {
        val nvgInstance = nvg
        val accentColor = accent
        val palette = palette

        val x = getX()
        val y = getY()
        val width = getWidth()
        val height = getHeight()
        val circle = 11 * scale
        val toggled = setting.isToggled()

        opacityAnimation.setAnimation(if (toggled) 1.0f else 0.0f, 14.0)
        toggleAnimation.setAnimation(if (toggled) 20.5f else 2.5f, 14.0)

        nvgInstance.drawRoundedRect(x, y, width, height, 7 * scale, palette.getBackgroundColor(ColorType.NORMAL))
        nvgInstance.drawGradientRoundedRect(
            x,
            y,
            width,
            height,
            7.4f * scale,
            ColorUtils.applyAlpha(accentColor.color1, (opacityAnimation.value * 255).toInt()),
            ColorUtils.applyAlpha(accentColor.color2, (opacityAnimation.value * 255).toInt())
        )
        nvgInstance.drawRoundedRect(
            x + toggleAnimation.value * scale,
            y + 2.5f * scale,
            circle,
            circle,
            circle / 2,
            circleAnimation.getColor(if (toggled) Color.WHITE else palette.getBackgroundColor(ColorType.DARK), 16)
        )

        super.draw(mouseX, mouseY, partialTicks)
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        val x = getX()
        val y = getY()
        val width = getWidth()
        val height = getHeight()

        if (MouseUtils.isInside(mouseX, mouseY, x, y, width, height) && mouseButton == 0) {
            setting.setToggled(!setting.isToggled())
        }

        super.mouseClicked(mouseX, mouseY, mouseButton)
    }
}
